05_testing_and_ci/automation_exercise.md

```diff
@@ -0,0 +1,71 @@
+- Install GitLab Runner on your machine and register a runner for your repository/fork. Ideally, it should be a runner that can work with the `docker` executor and is able to run Linux images. You can find the [detailed installation instructions](https://docs.gitlab.com/runner/install/). It is usually easiest to run install GitLab Runner as [Docker container](https://docs.gitlab.com/runner/install/docker.html).
+
+  If you cannot add such a runner on your machine, please sent an email with the registration token and your GitLab username to [alexander.jaust@ipvs.uni-stuttgart.de](mailto:alexander.jaust@ipvs.uni-stuttgart.de). We will add a runner to your repository/fork.
+
+- Set up the GitLab pipeline to run the checks, tests, and to create the coverage report.
+    - The pipeline should have three jobs (`style_check`, `test`, `coverage_report`). Each of these jobs should carry out the corresponding step described in [Task descriptions](## Task descriptions).
+    - The intermediate coverage information (`.coverage`) should be kept for [three hours](https://docs.gitlab.com/ee/ci/yaml/index.html#artifactsexpire_in).
+    - Make sure that your runner is online while you make changes to the repository. Otherwise the CI/CD pipeline cannot run.
+    - You may use the Docker image `ajaust/automation-exercise` (based on Ubuntu). You can find its Dockerfile in the GitLab repository (TODO: Add Dockerfile.).
+    - If a pipeline does not finish, check the output of its jobs and cancel it manually if necessary.
+
+- Add a [GitLab pipeline status badge](https://docs.gitlab.com/ee/user/project/badges.html) to the project that is based on your pipeline and the main branch. You do not have to add the badge to the Readme as it is added automatically to the repository.
+    - Label this badge with "SSE CI/CD".
+    - Follow the suggestions for "Link" and "Badge image URL", but replace the `https://gitlab.com` by `https://gitlab-sim.informatik.uni-stuttgart.de`.
+- Submit your solution via amerge request. If you are opening the merge request before you are finished, please mark it as draft.
+    - Create an issue in the [automation exercise repository (TODO: Update link)](https://gitlab-sim.informatik.uni-stuttgart.de/sse-test-group/automation-exercise).
+    - The issue should be named `[USERNAME] GitLab CI/CD exercise`, e.g., `[jaustar] GitLab CI/CD exercise`. Please use the GitLab username here.
+
+## Optional tasks
+
+- TODO Coverage visualization: Once the pipeline works as intended, we can add coverage report visualization via the [artifacts reports feature](https://docs.gitlab.com/ee/ci/yaml/index.html#artifactsreports). For the XML file generated by `coverage` this feature is [integrated in GitLab](https://docs.gitlab.com/ee/user/project/merge_requests/test_coverage_visualization.html#python-example).
+
+## Hints and remarks
+
+TODO
```

Comment:
```
jobs style job config should implement a script: or a trigger: keyword
```
confused me for a while. Here, a hint could be necessary.

Also the default OS of the runner confused me for a while. I guess this could be clearer after your demo though.

Reply:
What do you mean with OS? The note that the runner must be able to run Linux containers?

Reply:
No, that in the registration dialogue (of a runner), one has to specify a default OS. But you explained this in the lecture :+1:

Reply:
Great. I will also extend the demo notes a bit to make this a bit more clear.